Multiline comments

little-bobby-tables opened the pull request:
I've noticed that Slim handles HTML comments (`/!`)  just like verbatim text nodes (`|` and `'`) — complete with interpolation — so I decided to move the common semantics into a `text_block` rule.

The change introduces:
* multiline HTML comments (`/!`)
* text interpolation in HTML comments
* mutlilne code comments (`/`)

What do you think?

little-bobby-tables committed:
initial text_block implementation

little-bobby-tables committed:
multiline comments

little-bobby-tables committed:
dry the code up a bit

Rakoth commented on lib/slime/parser/text_block.ex:
```diff
@@ -0,0 +1,46 @@
+   ^
+  declaration indent
+  """
+  def render(lines, decl_indent, trailing_whitespace \\ "") do
```
May be use `declaration_indent` as variable name? An abbreviation is a bit confusing.

Rakoth commented on lib/slime/parser/text_block.ex:
```diff
@@ -0,0 +1,46 @@
+    [{first_line_indent, first_line, is_eex_line} | rest] = lines
+
+    text_indent = if first_line == "" do
+      [{indent, _, _} | _] = rest
```
what about single `'` used for spacing, example similar to one in my work project:
```
p
  '
  span text
  ' .
```
I understand, `span<` could be used here, but it is a valid slim and should be supported (probably should add a test case for it)

Rakoth commented on src/slime_parser.peg.eex:
```diff
@@ -8,10 +8,10 @@ doctype <- space? 'doctype' space (!eol .)+ eol;
-% NOTE: We have to track leading spaces for verbatim_text, so it handled separately
-tag <- verbatim_text / space? tag_item / blank:(space? &eol);
+% NOTE: Leading whitespace is handled separately for text items
+tag <- text_item / space? tag_item / blank:(space? &eol);
```
I think it is better to keep `verbatim_text` separated from `comment`, so here it would be `comment / verbatim_text / ...`

Rakoth reviewed (changes requested):
Thanks, @little-bobby-tables, great job refactoring this nasty verbatim_text transform! I left some comments below, please take a look.

little-bobby-tables committed:
fix text block indentation rendering

little-bobby-tables committed:
split text_item into verbatim_text and comment

little-bobby-tables commented:
Thanks for your comments, @Rakoth! I've added a test case for `'` without content and split `text_item` into `verbatim_text` and `comment`.

However, I've noticed that credo now reports the TextBlock.render function as having excessively high ABC complexity — do you have any ideas on how to refactor it?

Rakoth commented:
@little-bobby-tables  Since it is already in a separate module, we could simply extract parts of it into private functions.

little-bobby-tables committed:
move line spacing insertion into a separate func

Rakoth reviewed (approved):
LGTM 👍
